packages/dialog/src/mwc-dialog-base.ts

```diff
@@ -118,7 +120,58 @@ export class DialogBase extends BaseElement {
   }
 
   protected getInitialFocusEl(): HTMLElement|null {
-    return this.querySelector(`[${this.initialFocusAttribute}]`);
+    const initFocusSelector = `[${this.initialFocusAttribute}]`;
+
+    // only search light DOM. This typically handles all the cases
+    const lightDomQs = this.querySelector(initFocusSelector);
+
+    if (lightDomQs) {
+      return lightDomQs as HTMLElement;
+    }
+
+    // if not in light dom, search each flattened distributed node.
+    const content = this.contentSlot as HTMLSlotElement;
+    const primary = this.primarySlot as HTMLSlotElement;
+    const secondary = this.secondarySlot as HTMLSlotElement;
+
+    const contentNodes = content.assignedNodes({flatten: true});
+    const primaryNodes = primary.assignedNodes({flatten: true});
+    const secondaryNodes = secondary.assignedNodes({flatten: true});
+
+    const initFocusElement = this.searchNodeTreesForAttribute(
+        contentNodes, this.initialFocusAttribute);
+    if (initFocusElement) {
+      return initFocusElement;
+    }
+
+    const primaryFocusElement = this.searchNodeTreesForAttribute(
+        primaryNodes, this.initialFocusAttribute);
+    if (primaryFocusElement) {
+      return primaryFocusElement;
+    }
+    const secondaryFocusElement = this.searchNodeTreesForAttribute(
+        secondaryNodes, this.initialFocusAttribute);
+    return secondaryFocusElement;
+  }
+
+  private searchNodeTreesForAttribute(nodes: Node[], attribute: string):
+      HTMLElement|null {
+    for (const node of nodes) {
+      if (!(node instanceof HTMLElement)) {
+        continue;
+      }
+
+      if (node.hasAttribute(attribute)) {
+        return node;
+      } else {
+        const selection = node.querySelector(`[${attribute}]`);
+        if (selection instanceof HTMLElement) {
```

Comment:
Is it possible for these nodes not to be HTMLElements?

Reply:
`querySelector` returns `null|Node`. So this both checks whether it's null and casts it to HTMLElement at the same time

Reply:
Yeah, but the instanceof check happens at runtime so there is a cost vs a cast. Is there something that could have the attribute but not be an HTMLElement (genuine question -- an SVG element or something? would that even work?)

Reply:
good point